modules/clients/src/test/java/org/apache/ignite/jdbc/thin/JdbcThinConnectionFailoverTest.java

```diff
@@ -0,0 +1,252 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.jdbc.thin;
+
+import java.sql.Connection;
+import java.sql.DriverManager;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Statement;
+import org.apache.ignite.configuration.ClientConnectorConfiguration;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.junit.Test;
+
+import static org.junit.Assert.assertArrayEquals;
+
+/**
+ * JDBC driver transparent reconnection test with single address.
+ */
+public class JdbcThinConnectionFailoverTest extends JdbcThinAbstractSelfTest {
+    /** Nodes count. */
+    private static final int NODES_CNT = 2;
+
+    /** */
+    private static final String STATEMENT_CLOSED_ON_DISCONNECT_MESSAGE =
+        "Statement was closed due to a disconnection from the server";
+
+    /** */
+    private static final String CONNECT_URL = "jdbc:ignite:thin://127.0.0.1:"
+        + ClientConnectorConfiguration.DFLT_PORT;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String name) throws Exception {
+        return super.getConfiguration(name)
+            .setClientConnectorConfiguration(new ClientConnectorConfiguration());
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void beforeTest() throws Exception {
+        super.beforeTest();
+
+        startGrids(NODES_CNT);
+    }
+
+    /** {@inheritDoc} */
+    @Override protected void afterTest() throws Exception {
+        stopAllGrids();
+    }
+
+    @Test
+    public void testReconnectAfterRestart() throws Exception {
+        try (Connection conn = DriverManager.getConnection(CONNECT_URL)) {
+            Statement stmt1 = createTestTable(conn);
+
+            {
+                restartGrid(0);
+
+                Statement stmt = conn.createStatement();
+
+                assertEquals(1, stmt.executeUpdate("INSERT INTO TEST VALUES (1, 1)"));
+                stmt.execute("SELECT 1");
+            }
+
+            {
+                PreparedStatement pstmt = conn.prepareStatement("INSERT INTO TEST VALUES (?, ?)");
+
+                pstmt.setInt(1, 2);
+                pstmt.setInt(2, 2);
+
+                restartGrid(0);
+
+                assertEquals(1, pstmt.executeUpdate());
+
+                restartGrid(0);
+
+                pstmt.setInt(1, 3);
+                pstmt.setInt(2, 3);
+                pstmt.addBatch();
+
+                // Batch cannot be re-used after connection reset.
+                assertThrowsSql(pstmt::executeBatch, "Failed to communicate with Ignite cluster");
+                assertThrowsSql(() -> pstmt.setInt(1, 5), STATEMENT_CLOSED_ON_DISCONNECT_MESSAGE);
+
+                restartGrid(0);
+
+                PreparedStatement pstmt0 = conn.prepareStatement("INSERT INTO TEST VALUES (?, ?)");
```

Comment:
seems lost check  for the prepared statement after restart

Reply:
PR reworked to re-establish connection only on idempotent requests. These tests reworked accordingly.